apis/v1alpha1/backendpolicy_types.go

```diff
@@ -0,0 +1,173 @@
+	// following to specify a service:
+	//
+	// fooRef:
+	//   resource: services
+	//   name: myservice
+	//
+	// Otherwise, if the core API group is not desired, specify the desired
+	// group:
+	//
+	// fooRef:
+	//   group: acme.io
+	//   resource: foos
+	//   name: myfoo
+	//
+	// +optional
+	// +kubebuilder:default=core
+	Group string `json:"group"`
+
+	// Resource is the API resource name of the referent. Omitting the value
+	// or specifying the empty string indicates the services resource. For example,
+	// use the following to specify a services resource:
+	//
+	// fooRef:
+	//   name: myservice
+	//
+	// Otherwise, if the services resource is not desired, specify the desired
+	// group:
+	//
+	// fooRef:
+	//   group: acme.io
+	//   resource: foos
+	//   name: myfoo
+	//
+	// +optional
+	// +kubebuilder:default=services
+	Resource string `json:"resource"`
+
+	// Name is the name of the referent.
+	//
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+
+	// Port is the port of the referent. If unspecified, this policy applies to
+	// all ports on the backend.
+	// +optional
+	Port *int32 `json:"port,omitempty"`
+}
+
+// BackendTLSConfig describes TLS configuration for a backend.
+type BackendTLSConfig struct {
+	// ClientCertificateRef is a reference to a TLS client certificate that may
+	// be used to connect to these backends. If an entry in this list omits or
+	// specifies the empty string for both the group and the resource, the
+	// resource defaults to "secrets". An implementation may support other
+	// resources (for example, resource "mycertificates" in group
+	// "networking.acme.io").
+	//
+	// If a Secret is referenced, it must be of type "kubernetes.io/tls" and
+	// contain tls.crt and tls.key data fields that contain the certificate and
+	// private key to use for TLS.
+	//
+	// Support: Extended
+	//
+	// +optional
+	ClientCertificateRef *CertificateObjectReference `json:"clientCertificateRef,omitempty"`
+
+	// CertificateAuthorityRef is a reference to a resource that includes
+	// trusted CA certificates for the associated backends. If an entry in this
+	// list omits or specifies the empty string for both the group and the
+	// resource, the resource defaults to "secrets". An implementation may
+	// support other resources (for example, resource "mycertificates" in group
+	// "networking.acme.io").
+	//
+	// Support: Extended
+	//
+	// +optional
+	CertificateAuthorityRef *CertificateObjectReference `json:"certificateAuthorityRef,omitempty"`
+
+	// Options are a list of key/value pairs to give extended options to the
+	// provider.
+	//
+	// Support: Implementation-specific.
+	// +optional
+	Options map[string]string `json:"options,omitempty"`
+}
+
+// BackendPolicyStatus defines the observed state of BackendPolicy.
+type BackendPolicyStatus struct {
+	// Conditions describe the current conditions of the BackendPolicy.
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
```

Comment:
We can do this in future, but should we structure status subresource in a way that allows for reporting status at a per backendRef granularity?

Reply:
I could be wrong here, but I think this may the wrong direction. When would a timeout be valid for one backend and not another? Maybe we need a way of indicating that a backendRef does not resolve to a resource? What might be more significant here is if more than one GatewayClass/controller is using a backendRef and a field on that backendRef is invalid for them. Each controller may need a way to indicate that separately. Not quite sure if we've found a great way to solve this yet, but it feels like a similar problem to Route status...

Reply:
AppProtocol may be a per-backend issue. Problems could appear on route status instead?

Reply:
Good point. I was thinking more about the happy path where user wants feedback on whether this BackendPolicy was successfully applied to which services.

Reply:
Since there are so many potential ways to slice this up, I think the simplest for now is to have per Backend conditions at this level and then place implementation specific status on Routes (which already have per-Gateway status). I've updated the comments to suggest that approach here.

Reply:
> implementation specific status on Routes

Umm, to be clear, the policy is suppose to target a resource like core.Service and not a Route, right?
The whole point of #196 is to associate such configuration, in _most_ cases, to upstream Services and not with Routes.


(I'm fine with the status being what it is here but want to clarify how we are thinking about using backendPolicy)

Reply:
Yep, absolutely. This was just intended as a way to avoid a large number of dimensions in BackendPolicyStatus. As an example, we might end up with a top level set of conditions here and then conditions per backend + conditions per route targeting this backend. This at least temporarily shifts that last one to the Route. I tried to clean up the comments here so it's a bit clearer.